File: core/ui/api/src/androidMain/kotlin/com/isharaw/kmpproj/core/CapabilityGate.kt
package com.isharaw.kmpproj.core

import androidx.compose.runtime.Composable

/**
 * **Android (Compose) capability wrapper.** Renders [content] only if [snapshot] grants [capability]
 * (a dotted key, e.g. `"order.create"`); otherwise renders [fallback] (nothing by default). The
 * matching iOS wrapper is `CapabilityGate` in `iosApp` (SwiftUI) — both share
 * `ExperienceSnapshot.hasCapability`.
 *
 * ```
 * CapabilityGate(snapshot, "order.create") {
 *     Button(onClick = { … }) { Text("New order") }
 * }
 * ```
 */
@Composable
fun CapabilityGate(
    snapshot: ExperienceSnapshot,
    capability: String,
    fallback: @Composable () -> Unit = {},
    content: @Composable () -> Unit,
) {
    if (snapshot.hasCapability(capability)) content() else fallback()
}

File: iosApp/iosApp/CapabilityGate.swift
import SwiftUI
import Shared

/// **iOS (SwiftUI) capability wrapper** — the counterpart of the Android Compose `CapabilityGate`.
/// Renders `content` only if `snapshot` grants `capability` (a dotted key, e.g. "order.create").
///
/// Both platforms share the same Kotlin check: `ExperienceSnapshot.hasCapability` (now an O(1) member
/// on the snapshot, exported from the `Shared` framework).
///
/// ```swift
/// CapabilityGate(snapshot: snapshot, capability: "order.create") {
///     Button("New order") { … }
/// }
/// ```
struct CapabilityGate<Content: View>: View {
    let snapshot: ExperienceSnapshot
    let capability: String
    @ViewBuilder let content: () -> Content

    var body: some View {
        if snapshot.hasCapability(capability: capability) {
            content()
        }
    }
}
